cmd/tailscale/cli/serve_test.go

```diff
@@ -631,23 +617,58 @@ func TestServeConfigMutations(t *testing.T) {
 			continue
 		}
 		if st.wantErr != nil {
-			t.Fatalf("step #%d, line %v: got success (saved=%v), but wanted an error", i, st.line, newState != nil)
+			t.Fatalf("step #%d, line %v: got success (saved=%v), but wanted an error", i, st.line, lc.newState != nil)
 		}
-		if !reflect.DeepEqual(newState, st.want) {
+		if !reflect.DeepEqual(lc.newState, st.want) {
 			t.Fatalf("[%d] %v: bad state. got:\n%s\n\nwant:\n%s\n",
-				i, st.command, asJSON(newState), asJSON(st.want))
+				i, st.command, asJSON(lc.newState), asJSON(st.want))
 			// NOTE: asJSON will omit empty fields, which might make
 			// result in bad state got/want diffs being the same, even
 			// though the actual state is different. Use below to debug:
 			// t.Fatalf("[%d] %v: bad state. got:\n%+v\n\nwant:\n%+v\n",
-			// i, st.command, newState, st.want)
+			// i, st.command, lc.newState, st.want)
 		}
-		if newState != nil {
-			current = newState
+		if lc.newState != nil {
+			lc.current = lc.newState
 		}
 	}
 }
 
+// fakeLocalServeClient is a fake implementation of LocalClient
+// used by the serve command tests. It's not a full implementation,
+// just enough to test the serve command.
+//
+// The fake client is stateful, and is used to test manipulating
+// ServeConfig state.
+type fakeLocalServeClient struct {
+	current  *ipn.ServeConfig
+	newState *ipn.ServeConfig
```

Comment:
yeah, why track the values differently?

Reply:
Fine to simplify and change to a single field. But to answer your questions:

Two fields because 1) the way it was initially written; 2) the tests check their `want` against config passed to `SetServeConfig(...)` during a given step.

Example:
https://github.com/tailscale/tailscale/blob/4073613ffe18a7000825ae5501e083ac53fe9c93/cmd/tailscale/cli/serve_test.go#L69-L76

If we switch to one field then the step on L73 want would be an unchanged `ServeConfig` from the previous step, not nil:
```
{
    "AllowFunnel": {
        "foo.test.ts.net:443": true
    }
}
```

Reply:
Yeah, but the old code was not trying to implement any interface. It was defining exactly what it needed for tests: a `reset` bool to clear the state if needed for that series of stateful steps, `newState` to capture what was saved (and whether a save even occured!), and `current`, to keep track of the accumulated state between steps when there isn't a `reset`.

I think you want to keep the `var newState *ipn.ServeConfig` line above unchanged so you can tell whether a save happened.

Reply:
PTAL